src/cls/IPM/General/History.cls

```diff
@@ -0,0 +1,394 @@
+        $ListBuild("Package", $$$Green),
+        $ListBuild("Version", $$$Green),
+        $ListBuild("UserName", $$$Blue),
+        $ListBuild("Namespace", $$$Magenta),
+        $ListBuild("Time", $$$Yellow),
+        $ListBuild("Success", $$$Default),
+        $ListBuild("Committed", $$$Cyan),
+        $ListBuild("CommandString", $$$Red),
+        $ListBuild("Source", $$$Default)
+    )
+
+    Set ptr = 0
+    Write !
+    While $ListNext(columnStyles, ptr, pair) {
+        Set $ListBuild(column, style) = pair
+        If detail = "header" {
+            Write $$$FormattedLine(style, column), " "
+            Continue
+        }
+        Set cls = $classname()
+        Set mthd = column_"ToString"
+        If $System.CLS.IsMthd(cls, mthd) {
+            Set value = $classmethod(cls, mthd, rs, detail)
+        } Else {
+            Set value = rs.%Get(column)
+        }
+        Set value = $Replace(value, $Char(13), $$$FormattedLine($$$Dim, "<CR>"))
+        Set value = $Replace(value, $Char(10), $$$FormattedLine($$$Dim, "<LF>"))
+        Set value = $Replace(value, $Char(9), $$$FormattedLine($$$Dim, "<TAB>"))
+        Set value = $$$FormattedLine(style, value)
+        If detail {
+            Write $$$FormattedLine($$$Dim, column), ": ", value, !
+        } Else { 
+            Write value, " " 
+        }
+    }
+    Return $$$OK
+}
+
+ClassMethod VersionToString(rs As %SQL.StatementResult, detail As %Boolean = 0) As %String [ Internal, Private ]
+{
+    Set obj = ##class(%IPM.General.History).%OpenId(rs.%Get("ID"))
+    If $IsObject(obj) {
+        Set str = obj.Version.ToString()
+        If str '= "0.0.0-0" {
+            Quit "v"_str
+        }
+    }
+    Quit "<unset>"
+}
+
+ClassMethod TimeToString(rs As %SQL.StatementResult, detail As %Boolean = 0) As %String [ Internal, Private ]
+{
+    If detail {
+        Quit "Started At " _ rs.%Get("TimeStart") _ "; Ended At " _ rs.%Get("TimeEnd")
+    } 
+
+    Set start = rs.%Get("TimeStart")
+    Set start = $Zdatetimeh(start, 3)
+    set $ListBuild(startDate, startTime) = $ListFromString(start, ",")
+    Set startTimestamp = startDate * 86400 + startTime
+    Set end = rs.%Get("TimeEnd")
+    If end = "" {
+        Quit rs.%Get("TimeStart")
+    } 
+
+    Set end = $Zdatetimeh(end, 3)
+    Set $ListBuild(endDate, endTime) = $ListFromString(end, ",")
+    Set endTimeStamp = endDate * 86400 + endTime
+    Set diff =  endTime - startTime
+
+    Set diffDays = diff \ 86400
+    Set diff = diff # 86400
+    Set diffHours = diff \ 3600
+    Set diff = diff # 3600
+    Set diffMinutes = diff \ 60
+    Set diffSeconds = diff # 60
+
+    Set duration = $Select(diffDays: diffDays _ "D ", 1:"") 
+        _$Select(diffHours: diffHours _ "h ", 1:"")
+        _$Select(diffMinutes: diffMinutes _ "m ", 1:"")
+        _$Select(diffSeconds: diffSeconds _ "s ", 1:"")
+
+    If duration = "" {
+        Set duration = "0s"
+    }
+
+    Quit rs.%Get("TimeStart") _ " (" _ $Zstrip(duration, "<>W") _ ")"
```

Comment:
I'd wonder about using $System.SQL.DATEDIFF and just always using seconds as the unit. Ideally this will always be a manageably small number.